### PR TITLE
Don't capture AsyncLocals into EventCounter timer

### PR DESCRIPTION
**Pull request by benaadams**
Resolves: https://github.com/dotnet/corefx/issues/26074

### CODE REVIEW

**Comment by karelz** [comment]
@brianrob @vancem can you please take a look?

**Comment by karelz** [comment]
@brianrob @vancem ping?

File: src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs [hunk 1]
```diff
@@ -407,7 +407,24 @@ private void EnableTimer(float pollingIntervalInSeconds)
...
+                bool restoreFlow = false;
+                try
+                {
+                    if (!ExecutionContext.IsFlowSuppressed())
+                    {
+                        ExecutionContext.SuppressFlow();
+                        restoreFlow = true;
+                    }
+
+                    _pollingTimer = new Timer(OnTimer, null, _pollingIntervalInMilliseconds, _pollingIntervalInMilliseconds);
```
Comment (stephentoub) [inline src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs]:
```C#
new Timer(s => ((EventCounter)s).OnTimer(null), this, _pollingIntervalInMilliseconds, _pollingIntervalInMilliseconds);
```

**Comment by stephentoub** [comment]
The OnTimer that's being invoked makes calls to _eventSource.Write.  That Write call will in turn invoke user code from any registered EventListener.  Previously that would have been done in the context of the captured ExecutionContext; now it won't.

**Comment by vancem** [comment]
To address Stephen's concern.   It is true that this changes the executioncontext that EventListeners will see when EventCounter events are fired.   However, and EventCounter event represents an aggregation of many things and thus its context is not really valuable (at it only represents the current thread).  In addition the timer is started by the thread that turns on the eventCounter.   For ETW this is an INJECTED thread (not likely to be useful), and for EventListeners, it is the thread that called the 'Enable' API (that is a thread that logically belongs to the monitoring infrastructure).    This makes it even less likely to be useful.

Thus I am OK with the semantics of this change.

File: src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs (hunk 1 above)
Comment (benaadams) [reply src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs]: Changed

**Comment by benaadams** [comment]
System.Reflection.Metadata.Tests.MetadataReaderTests/ RedHat.73.Amd64.Open-Release-x64
```
System.BadImageFormatException : Invalid COR20 header signature.
```

**Comment by benaadams** [comment]
test Linux x64 Release Build

**Comment by vancem** [comment]
@dotnet-bot  test NETFX x86 Release Build

**Pull request merged**